apps/aecore/src/aec_peers.erl

```diff
@@ -655,43 +608,78 @@ ping_interval_limits() ->
         true ->
-            lager:debug("will not ping ~p", [Uri]),
+            lager:debug("will not ping ~p", [PeerId]),
             ignore
     end.
 
-is_local_uri(Peer, #state{local_peer = LocalPeer}) ->
-    R = Peer#peer.port == LocalPeer#peer.port andalso 
-        lists:member(Peer#peer.host, [<<"localhost">>, <<"127.0.0.1">>, LocalPeer#peer.host]),
-    lager:debug("is_local_uri(~p) -> ~p (~p)", [uri_of_peer(Peer), R, uri_of_peer(LocalPeer)]),
+maybe_retry_peer(P = #peer{ connection = {error, PeerCon} }) ->
+    lager:debug("will retry peer ~p", [peer_info(P)]),
+    aec_peer_connection:retry(PeerCon);
+maybe_retry_peer(P) ->
+    lager:debug("will not retry peer ~p", [peer_info(P)]),
+    ok.
+
+is_local(Peer, #state{local_peer = LocalPeer}) ->
+    #{ pubkey := PubKey1 } = peer_info(Peer),
+    #{ pubkey := PubKey2 } = peer_info(LocalPeer),
+    R = PubKey1 == PubKey2,
+    lager:debug("is_local_uri(~p) -> ~p (~p)", [Peer, R, LocalPeer]),
     R.
 
--spec parse_uri(http_uri_uri()) -> peer() | {error, any()}.
-parse_uri(Uri) ->
-    case uri_safe_parse(Uri) of
-        {ok, {Scheme, _UserInfo, Host, Port, _Path, _Query, _Fragment}} ->
-            #peer{scheme = Scheme, host = iolist_to_binary(Host), port = Port};
-        {ok, {Scheme, _UserInfo, Host, Port, _Path, _Query}} ->
-            #peer{scheme = Scheme, host = iolist_to_binary(Host), port = Port};
-        {error, _} = Error ->
-            Error
+peer_id(PeerId) when is_binary(PeerId) ->
+    PeerId;
+peer_id(#{ host := Host, port := Port, pubkey := PubKey }) ->
+    peer_id(PubKey, Host, Port);
+peer_id(#peer{ host = Host, port = Port, pubkey = PubKey }) ->
+    peer_id(PubKey, Host, Port).
+
+peer_info(PeerId) when is_binary(PeerId) ->
+    split_peer_id(PeerId);
+peer_info(#peer{ host = H, port = P, pubkey = PK }) ->
+    #{ host => H, port => P, pubkey => PK };
+peer_info(PeerInfo = #{ host := _, port := _, pubkey := _ }) ->
+    PeerInfo.
+
+peer_id(PubKey, Host, Port) when is_binary(Host) ->
+    <<PubKey/binary, Port:16, Host/binary>>;
+peer_id(PubKey, Host, Port) when is_list(Host) ->
+    peer_id(PubKey, list_to_binary(Host), Port).
+
+split_peer_id(<<PubKey:32/binary, Port:16, Host/binary>>) ->
+    #{ pubkey => PubKey, host => Host, port => Port }.
+
+set_retry_timeout(Peer, Timeout) ->
+    set_timeout(Peer, Timeout, retry).
+
+set_ping_timeout(Peer, Timeout) ->
+    set_timeout(Peer, Timeout, ping).
+
+set_timeout(Peer = #peer{ timer_tref = Prev }, Timeout, Msg) ->
+    lager:debug("Starting ~p timer for ~p: ~p", [Msg, peer_info(Peer), Timeout]),
+    TRef = erlang:start_timer(Timeout, self(), {Msg, peer_id(Peer)}),
+    case Prev of
+        undefined -> ok;
+        _ -> try erlang:cancel_timer(Prev, [{async, true}, {info, false}])
+             catch error:_ -> ok end
+    end,
+    Peer#peer{timer_tref = TRef}.
+
+backoff_timeout(#peer{ retries = Retries, trusted = Trusted }) ->
+    case Retries of
+        N when N < ?MAX_RETRIES ->
```

Comment:
How do you keep MAX_RETRIES and this list in sync?
Perhaps define BACKOFF_TIMES at the same place as MAX_RETRIES is defined?

Reply:
Good comment - will fix!